sklearn/cluster/_kmeans.py

```diff
@@ -863,6 +864,37 @@ def _check_test_data(self, X):
 
         return X
 
+    def _check_mkl_vcomp(self, X, n_samples):
+        """Warns when vcomp and mkl are both present"""
+        # The BLAS call inside a prange in lloyd_iter_chunked_dense is known to
+        # cause a small memory leak when there are less chunks than the number
+        # of available threads. It only happens when the OpenMP library is
+        # vcomp (microsoft OpenMP) and the BLAS library is MKL.
+        if sp.issparse(X):
+            return
+
+        active_threads = int(np.ceil(n_samples / 256))
+        if active_threads < self._n_threads:
+            modules = threadpool_info()
+            has_vcomp = "vcomp" in [module["prefix"] for module in modules]
+            has_mkl = ("mkl", "intel") in [
+                (module["internal_api"], module.get("threading_layer", None))
+                for module in modules]
+            if has_vcomp and has_mkl:
+                if self.__class__ is KMeans:
+                    warnings.warn(
+                        f"KMeans is known to have a memory leak on Windows "
+                        f"with MKL. You can avoid it by setting the "
+                        f"environment variable "
+                        f"OMP_NUM_THREADS={active_threads}")
```

Comment:
In addition to giving the exact ideal value of OMP_NUM_THREADS I think it'd be good to explain where this number comes from, i.e say "when there are less chunks than the number of available threads"